packages/asset-swapper/src/network/utils.ts

```diff
@@ -0,0 +1,211 @@
+}
+
+export interface ContractWrapperType<T> {
+    new (
+        address: Address,
+        provider: SupportedProvider,
+        _txDefaults: {},
+        _logDeps: {},
+        deployedBytecode: Bytes | undefined,
+        encoderOverrides: {
+            encodeInput: (fnName: string, values: any) => Bytes;
+            decodeOutput: (fnName: string, data: Bytes) => any;
+        },
+    ): T;
+    ABI(): ContractAbi | MethodAbi[];
+}
+
+export function getDeterministicContractAddressFromBytecode(bytecode: Bytes): Address {
+    return hexUtils.leftPad(hexUtils.hash(bytecode), ADDRESS_SIZE);
+}
+
+export function getBytecodeFromArtifact(artifact: ContractVersionData): Bytes {
+    return artifact.compilerOutput.evm.deployedBytecode.object;
+}
+
+export function getDeterministicContractAddressFromArtifact(artifact: ContractVersionData): Address {
+    return getDeterministicContractAddressFromBytecode(getBytecodeFromArtifact(artifact));
+}
+
+export interface GeneratedContract extends BaseContract {
+    getABIDecodedReturnData: <T>(methodName: string, returnData: string) => T;
+}
+
+export type ContractFunction<TArgs extends any[], TReturn> = (...args: TArgs) => ContractFunctionObj<TReturn>;
+
+export type ValueByChainId<T> = { [k in ChainId]: T };
+
+// TODO(kimpers): Consolidate this implementation with the one in @0x/token-metadata
+export function valueByChainId<T>(rest: Partial<ValueByChainId<T>>, defaultValue: T): { [key in ChainId]: T } {
+    // TODO I don't like this but iterating through enums is weird
+    return {
+        [ChainId.Mainnet]: defaultValue,
+        [ChainId.Ropsten]: defaultValue,
+        [ChainId.Rinkeby]: defaultValue,
+        [ChainId.Kovan]: defaultValue,
+        [ChainId.Ganache]: defaultValue,
+        [ChainId.BSC]: defaultValue,
+        [ChainId.Polygon]: defaultValue,
+        [ChainId.PolygonMumbai]: defaultValue,
+        ...(rest || {}),
+    };
+}
+
+export function createContractWrapperAndHelper<TContract extends GeneratedContract>(
+    chain: Chain,
+    contractType: ContractWrapperType<TContract>,
+    artifactName: string,
+    address?: Address,
+): [TContract, ContractHelper<TContract>] {
+    const fastAbi = new FastABI(
+        contractType.ABI() as MethodAbi[],
+        { BigNumber },
+    );
+    const artifact = (artifacts as ArtifactsMap)[artifactName];
+    const wrapper = new contractType(
+        address || getDeterministicContractAddressFromArtifact(artifact),
+        DUMMY_PROVIDER,
+        {},
+        {},
+        artifact.compilerOutput.evm.deployedBytecode.object,
+        {
+            encodeInput: (fnName: string, values: any) => fastAbi.encodeInput(fnName, values),
+            decodeOutput: (fnName: string, data: string) => fastAbi.decodeOutput(fnName, data),
+        },
+    );
+    return [
+        wrapper,
+        new ContractHelper(chain, wrapper),
+    ];
+}
+
+export class ContractHelper<TBaseContract extends GeneratedContract> {
```

Comment:
Any time you want to do batching-compatible calls on a contract (wrapper), you should use it with a `ContractHelper` instance. It'll automatically include the correct overrides for the eth_call and makes things typesafe.

Reply:
Might want to put some nice jsdoc comment here so that it's easier to remember 😃

Reply:
But if you're looking at this class you're probably aware. The warning really needs to be near where someone would be tempted to call `callAsync()`. Maybe in `NetworkUtils` or `NativeOrderUtils`. For dealing with regular samplers it's all abstracted away atm.